### DISCUSSION TITLE
Can you use AutoValue with @JsonAny Getter ?

### DISCUSSION THREAD

**Discussion by shastrisubramaniam** [google/auto#1376] [Q&A]
Currenty when I use AutoValue with @JSonAnyGetter, I am getting this exception at Run time 

com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Invalid type definition for type `model.samples.OperationValue`: Argument #1 of constructor [method model.samples.OperationValue#create(2 params)] has no property name (and is not Injectable): can not use as property-based Creator

**Comment by eamonnmcmanus** [comment]
Could you include a code sample that reproduces the problem? (I think it may not be possible to do what you want to do, but I'm not sure.)

**Reply by shastrisubramaniam** [reply]
Following is the code I am using with AutoValue 

package model.samples;


import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonRawValue;
import com.google.auto.value.AutoValue;
import java.util.Map;

@AutoValue
public abstract class OperationValue {

    @JsonProperty("viridian_log")
    public abstract ViridianLog viridianLog();

    @JsonAnyGetter
    public abstract Map<String,SampleTree> sampleMap();
    
    @JsonCreator
    public static OperationValue create(@JsonProperty("viridian_log") ViridianLog viridianLog,Map<String,SampleTree> sampleMap) {
        return new AutoValue_OperationValue(viridianLog,sampleMap);
    }
}

I am trying to parse a section of the JSON where the key is primarily the GET parameter and deserialize it.

The error message when I run the code seems to indicate that the create method expects @JsonProperty params. Currently I have hardcoded the 2nd JsonProperty parameter, but I need to figure out how to pass that. Wondering if I need to create a custom DeSerializer to do that.

**Reply by tbroyer** [reply]
AIUI, `@JsonAnyGetter` is only for serialization; you'd need a `@JsonAnySetter` or equivalent for parsing.
Javadoc for `@JsonCreator` says every argument must be annotated with `@JsonProperty` or `@JacksonInject`, hence the error you're seeing; which I interpret as that there's no equivalent of `@JsonAnySetter` with a `@JsonCreator`, and you'd have to use builder or other custom deserializer.

I haven't tried but maybe something like this would work:
```java
@JsonDeserialize(builder = AutoValue_OperationValue.Builder.class)
@AutoValue
public abstract class OperationValue {
   // …

   @JsonPOJOBuilder(withPrefix = "set")
   @AutoValue.Builder
   static abstract class Builder {
      @JsonProperty("viridian_log")
      abstract Builder setViridianLog(ViridianLog log);

      abstract ImmutableMap.Builder<String, SampleTree> sampleMapBuilder();

      @JsonAnySetter
      final Builder addSampleTree(String name, SampleTree value) {
         sampleMapBuilder().add(name, value);
         return this;
      }

      abstract OperationValue build();
   }
}
```

Anyway, this is more a Jackson question (deserializing with JsonAnySetter or equivalent, with immutable value objects) than an AutoValue question; I think you'll have a better chance of getter an answer in the Jackson community.

**Reply by shastrisubramaniam** [reply]
Thank you !

**Reply by shastrisubramaniam** [reply]
I was able to get it working.
